carbon-design-system/carbon-addons-iot-react
# 3025 list add allowed drop target

Closes #3025

**Summary**
Adds the option to limit which rows can act as drop targets  using a callback getAllowedDropIds that receives the id of the item about to be dragged and should return an array of ids of the rows that are allowed as drop targets. I decided on using  a callback like this because providing some kind of map of AllowedDropIds up front would be too costly and using the existing `itemWillMove` wouldn't work since it was called after the drop targets were rendered.

**Change List (commits, features, bugs, etc)**
- Added prop `getAllowedDropIds` from component ListItemWrapper all the way up to the List and HierarchyList.
- Memoized `getAllowedDropIds` in the ListComponent to prevent a possibly expensive function to be executed for each row.
- Refactored some poorly written dnd code
- Created new stories for List and HierarchyList in order to have meaningful examples
- Added new unit tests
- Added new e2e tests
- Updated docs

**Acceptance Test (how to verify the PR)**
1. Go to https://deploy-preview-3039--carbon-addons-iot-react.netlify.app?path=/story/1-watson-iot-hierarchylist--with-nested-reordering-restricted
2. Make sure players can only be reordered within their team and not dropped in other teams
3. Make sure teams can only be reordered among other teams and not dropped in other teams 
4. Enable knob hasVirtualList and repeat tests

1. Go to https://deploy-preview-3039--carbon-addons-iot-react.netlify.app?path=/story/1-watson-iot-list--with-reorder-and-drop-target-restrictions
2. Make sure drops are only allowed on rows defined in knob `getAllowedDropIds return value`
3. Enable knob hasVirtualList and repeat tests

**Regression Test (how to make sure this PR doesn't break old functionality)**
- Make sure all other stories for List and HierarchyList works as before



**Things to look for during review**
